### PR TITLE
Remove broken CSs based on outer frame not matching thread's majority

### PR DESCRIPTION
**Pull request by dpallotti**
As discussed in http://b/162910324#comment5.
Bug: http://b/162910324

### CODE REVIEW

File: OrbitCore/CallstackData.cpp [hunk 1]
```diff
@@ -175,3 +176,58 @@ std::shared_ptr<CallStack> CallstackData::GetCallstackPtr(CallstackID callstack_
...
+    // Find the outer frame with the most occurrences.
+    uint64_t majority_outer_frame = 0;
+    uint64_t majority_outer_frame_count = 0;
+    for (const auto& outer_frame_and_count : count_by_outer_frame) {
+      if (outer_frame_and_count.second > majority_outer_frame_count) {
+        majority_outer_frame = outer_frame_and_count.first;
+        majority_outer_frame_count = outer_frame_and_count.second;
+      }
+    }
+    if (majority_outer_frame_count == 0) {
```
Comment (dimitry-) [inline OrbitCore/CallstackData.cpp:209]: nitpick: maybe check if it is empty before for loop?

**Comment by dpallotti** [comment]
Please have a look.

File: OrbitCore/CallstackData.cpp [hunk 2]
```diff
@@ -175,3 +176,58 @@ std::shared_ptr<CallStack> CallstackData::GetCallstackPtr(CallstackID callstack_
...
+      continue;
+    }
+
+    // Discard the CallstackEvents whose outer frame doesn't match the majority outer frame.
+    for (auto time_and_event_it = callstack_events.begin();
+         time_and_event_it != callstack_events.end();) {
+      const CallstackEvent& event = time_and_event_it->second;
+      const std::vector<uint64_t> frames =
```
Comment (dimitry-) [inline OrbitCore/CallstackData.cpp:217]: &

File: OrbitCore/CallstackData.cpp [hunk 3]
```diff
@@ -175,3 +176,58 @@ std::shared_ptr<CallStack> CallstackData::GetCallstackPtr(CallstackID callstack_
...
+    uint64_t majority_outer_frame_count = 0;
+    for (const auto& outer_frame_and_count : count_by_outer_frame) {
+      if (outer_frame_and_count.second > majority_outer_frame_count) {
+        majority_outer_frame = outer_frame_and_count.first;
+        majority_outer_frame_count = outer_frame_and_count.second;
+      }
+    }
+    if (majority_outer_frame_count == 0) {
+      continue;
+    }
```
Comment (dimitry-) [inline OrbitCore/CallstackData.cpp:211]: maybe also check for supermajority here? majority_outer_frame_count > 0.6*number of callstacks in the thread.

Just a safeguard for distributions like (21, 20, 20, 19, 20) where 21 is winning but something this picture does not look right.

File: OrbitCore/CallstackData.cpp [hunk 4]
```diff
@@ -175,3 +176,58 @@ std::shared_ptr<CallStack> CallstackData::GetCallstackPtr(CallstackID callstack_
...
+  uint32_t count_before_filtering = GetCallstackEventsCount();
+
+  for (auto& tid_and_events : callstack_events_by_tid_) {
+    std::map<uint64_t, CallstackEvent>& callstack_events = tid_and_events.second;
+
+    // Count the number of occurrences of each outer frame for this thread.
+    absl::flat_hash_map<uint64_t, uint64_t> count_by_outer_frame;
+    for (const auto& time_and_event : callstack_events) {
+      const CallstackEvent& event = time_and_event.second;
+      const std::vector<uint64_t> frames =
```
Comment (dimitry-) [inline OrbitCore/CallstackData.cpp:191]: &

**Review by dimitry- — COMMENTED** [review]
Also this function needs a test.

File: OrbitCore/CallstackData.cpp (hunk 4 above)
Comment (dpallotti) [reply OrbitCore/CallstackData.cpp:191]: Done.

File: OrbitCore/CallstackData.cpp (hunk 1 above)
Comment (dpallotti) [reply OrbitCore/CallstackData.cpp:209]: Done.

File: OrbitCore/CallstackData.cpp (hunk 3 above)
Comment (dpallotti) [reply OrbitCore/CallstackData.cpp:211]: What do you think the action should be in that case though? Should the filtering just not happen? Should all events be removed (that's a bit extreme)? Should the filtering stay the same but at least an error logged?

File: OrbitCore/CallstackData.cpp (hunk 2 above)
Comment (dpallotti) [reply OrbitCore/CallstackData.cpp:217]: Done.

**Review by dpallotti — COMMENTED** [review]
> Also this function needs a test.

You're right, on it.

File: OrbitCore/CallstackData.cpp (hunk 3 above)
Comment (dimitry-) [reply OrbitCore/CallstackData.cpp]: I think filtering should not be happening at this point - maybe print a warning/error message.

Comment (dpallotti) [reply OrbitCore/CallstackData.cpp:222]: Done, let me know what you think.

**Review by dimitry- — APPROVED** [review]

**Pull request merged**